pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java

```diff
@@ -1040,6 +1040,21 @@ public void markBatchMessagePublished() {
             .putSchemaIfAbsent(id, schema, schemaCompatibilityStrategy);
     }
 
+    @Override
+    public CompletableFuture<SchemaVersion> deleteSchema() {
+        String base = TopicName.get(getName()).getPartitionedTopicName();
+        String id = TopicName.get(base).getSchemaName();
+        SchemaRegistryService schemaRegistryService = brokerService.pulsar().getSchemaRegistryService();
+        return schemaRegistryService.getSchema(id)
+                .thenCompose(schema -> {
+                    if (schema != null) {
+                        return schemaRegistryService.deleteSchema(id, "");
```

Comment:
should we handle concurrent deletions?

Reply:
Thank you @sijie .

Reply:
@liketic 

1) I don't think we should change async to sync. since sync will be blocking threads, which can potentially lead to deadlock.

2) what I would suggest is to mask the exception thrown due to schema not exists when deleteSchema. so you logic would be something like below:

```
public CompletableFuture<SchemaVersion> deleteSchema() {
        String base = TopicName.get(getName()).getPartitionedTopicName();
        String id = TopicName.get(base).getSchemaName();
        SchemaRegistryService schemaRegistryService = brokerService.pulsar().getSchemaRegistryService();
        return schemaRegistryService.getSchema(id)
                .thenCompose(schema -> {
                    CompletableFuture<SchemaVersion> deleteFuture = new CompletableFuture<SchemaVersion>();
                    if (schema != null) {
                        schemaRegistryService.deleteSchema(id, "").whenComplete((version, cause) -> {
                            if (null != cause) {
                                if (cause instanceof SchemaNotFoundException) {
                                    deleteFuture.complete(null);
                                } else {
                                    deleteFuture.completeExceptionally(cause);
                                }
                            } else {
                                deleteFuture.complete(version);
                            }
                        });
                    } else {
                        deleteFuture.complete(null);
                    }
                    return deleteFuture;
                });
    }
```

Reply:
@sijie Thanks for your help! It's a good idea to check the result instead. However, I looked into the code of ```SchemaRegistryServiceImpl```, the delete operation only put a ```deleted``` entry to ```schemaStorage```, if we can accept ```deleteSchema``` been called more than once for the same schema id, we may don't need to care the concurrent deletion?

Reply:
@liketic ah you are right